### PR TITLE
porting shape optimization to pybind11

### PR DESCRIPTION
**Pull request by armingeiser**
The tests are running, except the ones including mesh motion.
@dbaumgaertner we need to have a look at this next week.

### CODE REVIEW

File: applications/ShapeOptimizationApplication/CMakeLists.txt [hunk 1]
```diff
@@ -14,14 +17,11 @@ set( KRATOS_SHAPEOPTIMIZATION_APPLICATION_SOURCES
 
 ###############################################################
 ## define library Kratos which defines the basic python interface
-add_library(KratosShapeOptimizationApplication SHARED ${KRATOS_SHAPEOPTIMIZATION_APPLICATION_SOURCES})
-target_link_libraries(KratosShapeOptimizationApplication KratosCore)
+pybind11_add_module(KratosShapeOptimizationApplication MODULE ${KRATOS_SHAPEOPTIMIZATION_APPLICATION_SOURCES})
+target_link_libraries(KratosShapeOptimizationApplication PRIVATE KratosCore)
```
Comment (RiccardoRossi) [inline applications/ShapeOptimizationApplication/CMakeLists.txt:21]: please add the "cotire" command

**Review by dbaumgaertner — COMMENTED** [review]
Looks good for me. Just a question. Boost::python is gone. But, without any special care, I can still use e.g. 

- boost::timer
- boost::hash_range
- boost::unordered_map
- or ublas includes like "boost/numeric/ublas/matrix.hpp", 

right?

**Comment by philbucher** [comment]
@dbaumgaertner 
- I suggest using the [builtin timer](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/builtin_timer.h) by @msandre 
- I think @adityaghantasala uses sth similar
- this is now in the stl: `std::unordered_map`
- I think (not sure though) that this should be included through the space, I guess it will be removed with the change to AMatrix

File: applications/ShapeOptimizationApplication/custom_utilities/response_functions/strain_energy_response_function.h [hunk 2]
```diff
@@ -207,16 +201,16 @@ class StrainEnergyResponseFunction : ResponseFunction
 	}
 
 	// --------------------------------------------------------------------------
-	boost::python::dict GetGradient()
```
Comment (RiccardoRossi) [inline applications/ShapeOptimizationApplication/custom_utilities/response_functions/strain_energy_response_function.h:210]: i am approving, however as a hint i would give back a std::unordered_map

that would help you insulating from python. We'll discuss this separately though

**Review by RiccardoRossi — APPROVED** [review]

**Comment by philbucher** [comment]
@loumalouomega I saw that you are doing sth with hashing.
Do you know if there is sth in Kratos to replace `boost::hash_range`?
Or how do you do your hashing?

File: applications/ShapeOptimizationApplication/custom_utilities/response_functions/strain_energy_response_function.h (hunk 2 above)
Comment (armingeiser) [reply applications/ShapeOptimizationApplication/custom_utilities/response_functions/strain_energy_response_function.h:210]: Thanks for the hint, this function is removed in PR #1804 . So i would keep this until we converged in the other, large PR.

**Comment by pooyan-dadvand** [comment]
I should see the use case but it is not very difficult to implement a `hash_range`

**Pull request merged**